File: packages/scout/src/cherrypick/scout/security.py
"""The settings-server invariant, ported to FastAPI middleware.

Scout is a research surface, not a settings editor, but it carries a handful of narrow mutating
routes (watchlist save, order dry-run, staged-ticket save/delete) — so it inherits the same posture
`cherrypick.orchestrator.settings_serve` uses for its one mutating HTTP surface, rather than the
plain loopback-binding-only rule every read-only dashboard in the suite gets away with: a malicious
webpage can fetch ``http://127.0.0.1:<port>`` from inside a user's browser, and DNS rebinding can
defeat same-origin, so binding alone is not enough once a route writes anything.

Every request (GET included) must carry a loopback ``Host`` header naming this exact port — else 403.
Every mutating request (POST/PUT/PATCH/DELETE) must additionally carry the per-process CSRF token
baked into the page, an ``application/json`` content type (which forces a cross-origin preflight this
server never answers, since it sends no CORS headers), and, when an ``Origin`` header is present, the
local origin.
"""

from __future__ import annotations

import hmac
import secrets as pysecrets

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse

_LOOPBACK_HOSTS = ("127.0.0.1", "localhost")
_MUTATING_METHODS = {"POST", "PUT", "PATCH", "DELETE"}

CSRF_HEADER = "x-csrf-token"


def new_csrf_token() -> str:
    return pysecrets.token_urlsafe(32)


class SecurityMiddleware(BaseHTTPMiddleware):
    """Loopback Host-header enforcement + CSRF/content-type/origin gating on mutating verbs.

    Constructed once per app with the bound port and the process's CSRF token, so a test can pass a
    fixed token and assert against it without reaching into global state.
    """

    def __init__(self, app, *, port: int, csrf_token: str):
        super().__init__(app)
        self._valid_hosts = {f"{h}:{port}" for h in _LOOPBACK_HOSTS} | set(_LOOPBACK_HOSTS)
        self._valid_origins = {f"http://{h}:{port}" for h in _LOOPBACK_HOSTS}
        self._csrf_token = csrf_token

    async def dispatch(self, request: Request, call_next):
        host = request.headers.get("host") or ""
        if host not in self._valid_hosts:
            return JSONResponse({"ok": False, "error": "bad Host header"}, status_code=403)

        if request.method in _MUTATING_METHODS:
            token = request.headers.get(CSRF_HEADER) or ""
            if not hmac.compare_digest(token, self._csrf_token):
                return JSONResponse({"ok": False, "error": "missing or invalid CSRF token"}, status_code=403)
            ctype = (request.headers.get("content-type") or "").split(";")[0].strip()
            if ctype != "application/json":
                return JSONResponse(
                    {"ok": False, "error": "Content-Type must be application/json"}, status_code=403
                )
            origin = request.headers.get("origin")
            if origin and origin not in self._valid_origins:
                return JSONResponse({"ok": False, "error": "cross-origin POST refused"}, status_code=403)

        return await call_next(request)

File: packages/console/server/src/security.ts
import crypto from "node:crypto";
import type { FastifyInstance, FastifyRequest, FastifyReply } from "fastify";

/**
 * The suite's mutating-surface posture (ported from scout/settings_serve):
 * loopback binding alone is not enough once a route writes anything — a
 * malicious webpage can fetch http://127.0.0.1:<port>, and DNS rebinding
 * defeats same-origin. Every request must present a loopback Host; every
 * mutating request additionally needs the per-process CSRF token, a JSON
 * content type, and a loopback Origin when one is present.
 */
export const CSRF_TOKEN = crypto.randomBytes(24).toString("hex");

const LOOPBACK_HOSTS = /^(127\.0\.0\.1|localhost|\[::1\])(:\d+)?$/;

function isLoopbackUrl(value: string): boolean {
  try {
    return LOOPBACK_HOSTS.test(new URL(value).host);
  } catch {
    return false;
  }
}

export function registerSecurity(app: FastifyInstance): void {
  app.addHook("onRequest", async (req: FastifyRequest, reply: FastifyReply) => {
    const host = req.headers.host ?? "";
    if (!LOOPBACK_HOSTS.test(host)) {
      return reply.code(403).send({ error: "forbidden host" });
    }
    if (req.method === "GET" || req.method === "HEAD" || req.method === "OPTIONS") return;

    const origin = req.headers.origin;
    if (origin !== undefined && !isLoopbackUrl(origin)) {
      return reply.code(403).send({ error: "forbidden origin" });
    }
    if (req.headers["x-csrf-token"] !== CSRF_TOKEN) {
      return reply.code(403).send({ error: "missing or invalid csrf token" });
    }
    const ct = req.headers["content-type"] ?? "";
    if (req.method !== "DELETE" && !ct.includes("application/json")) {
      return reply.code(415).send({ error: "json only" });
    }
  });

  // The SPA fetches the token once per session; Host-checked like everything else.
  app.get("/api/csrf", async () => ({ token: CSRF_TOKEN }));
}
